src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -0,0 +1,652 @@
+            archiveStream.Write(uidBytes);
+            archiveStream.Write(gidBytes);
+            archiveStream.Write(sizeBytes);
+            archiveStream.Write(mTimeBytes);
+            archiveStream.Write(checksumBytes);
+            archiveStream.WriteByte(typeFlagByte);
+            archiveStream.Write(linkNameBytes);
+
+            archiveStream.Write(magicBytes);
+            archiveStream.Write(versionBytes);
+            archiveStream.Write(uNameBytes);
+            archiveStream.Write(gNameBytes);
+            archiveStream.Write(devMajorBytes);
+            archiveStream.Write(devMinorBytes);
+
+            archiveStream.Write(prefixBytes);
+            archiveStream.Write(new byte[FieldLengths.PosixPadding]);
+
+            if (_dataStream != null)
+            {
+                WriteData(archiveStream, _dataStream);
+            }
+        }
+
+        // All formats save in the name byte array only the bytes that fit.
+        private int SaveNameFieldAsBytes(Span<byte> nameBytes, out byte[] fullNameBytes)
+        {
+            fullNameBytes = Encoding.ASCII.GetBytes(_name);
+            int nameBytesLength = Math.Min(fullNameBytes.Length, FieldLengths.Name);
+            int checksum = WriteLeftAlignedBytesAndGetChecksum(fullNameBytes.AsSpan(0, nameBytesLength), nameBytes);
+            return checksum;
+        }
+
+        // Ustar and PAX save in the name byte array only the bytes that fit, and the rest of the string  (the bytes that fit) get saved in the prefix byte array.
+        private int SavePosixNameFieldAsBytes(Span<byte> nameBytes, Span<byte> prefixBytes)
+        {
+            int checksum = SaveNameFieldAsBytes(nameBytes, out byte[] fullNameBytes);
+            if (fullNameBytes.Length > FieldLengths.Name)
+            {
+                int prefixBytesLength = Math.Min(fullNameBytes.Length - FieldLengths.Name, FieldLengths.Name);
+                checksum += WriteLeftAlignedBytesAndGetChecksum(fullNameBytes.AsSpan(FieldLengths.Name, prefixBytesLength), prefixBytes);
+            }
+            return checksum;
+        }
+
+        // Writes all the common fields shared by all formats into the specified spans.
+        private int SaveCommonFieldsAsBytes(Span<byte> _modeBytes, Span<byte> _uidBytes, Span<byte> _gidBytes, Span<byte> sizeBytes, Span<byte> _mTimeBytes, ref byte _typeFlagByte, Span<byte> _linkNameBytes)
+        {
+            byte[] modeBytes = TarHelpers.GetAsciiBytes(TarHelpers.ConvertDecimalToOctal(_mode));
+            int checksum = WriteRightAlignedBytesAndGetChecksum(modeBytes, _modeBytes);
+
+            byte[] uidBytes = TarHelpers.GetAsciiBytes(TarHelpers.ConvertDecimalToOctal(_uid));
+            checksum += WriteRightAlignedBytesAndGetChecksum(uidBytes, _uidBytes);
+
+            byte[] gidBytes = TarHelpers.GetAsciiBytes(TarHelpers.ConvertDecimalToOctal(_gid));
+            checksum += WriteRightAlignedBytesAndGetChecksum(gidBytes, _gidBytes);
```

Comment:
I'll stop commenting on these, too, but there are more efficient ways to do this rather than allocating new byte[]s each time.

Reply:
Fixed with ArrayPool.